macros/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+use quote::quote;
+
+extern crate proc_macro;
+
+const LOG_FILTER: &str = "RUST_LOG";
+
+// Sets environment variables from the cargo target's perspective
+//
+// These values need to come from the cargo target which can only be found
+// during macro expansion. Calling `env!('CARGO_PKG_NAME')` within
+// `telemetry_layer.rs` will be incorrect as the macro will have already
+// expanded leading to the constant value "fuel-telemetry" for all targets
+fn set_env_vars() -> proc_macro2::TokenStream {
+    quote! {
+        let mut exe_name = String::from("unknown");
+
+        if let Ok(exe) = std::env::current_exe() {
+            if let Some(name) = exe.file_name() {
+                if let Some(name_str) = name.to_str() {
+                    exe_name = name_str.to_string().replace(':', "_");
+                }
+            }
+        }
+
+        if std::env::var("TELEMETRY_PKG_NAME").is_err() {
+            std::env::set_var("TELEMETRY_PKG_NAME", exe_name);
+        }
+
+        if std::env::var("TELEMETRY_PKG_VERSION").is_err() {
+            std::env::set_var("TELEMETRY_PKG_VERSION", env!("CARGO_PKG_VERSION"));
+        }
+    }
+}
+
+// Starts the `FileWatcher` and `SystemInfoWatcher` daemon
+//
+// Warning: We need to create the `FileWatcher` and `SystemInfoWatcher`
+// before the `TelemetryLayer` as there is a race condition in the
+// thread runtime of `tracing` and the tokio runtime of `Reqwest`.
+// Swapping order of the two could lead to possible deadlocks.
+//
+// If the watchers fail to start, we silently ignore the errors as
+// telemetry should not impede the program from running.
+fn start_watchers() -> proc_macro2::TokenStream {
+    quote! {
+        // Start the `FileWatcher` and only care about fatal errors
+        let mut file_watcher = fuel_telemetry::file_watcher::FileWatcher::new();
+        if let Err(fuel_telemetry::errors::WatcherError::Fatal(_)) = file_watcher.start() {
+            std::process::exit(1);
+        }
+
+        // Start the `ProcessWatcher` and only care about fatal errors
+        let mut process_watcher = fuel_telemetry::process_watcher::ProcessWatcher::new();
+        if let Err(fuel_telemetry::errors::WatcherError::Fatal(_)) = process_watcher.start() {
+            std::process::exit(1);
+        }
+
+        // Start the `SystemInfoWatcher` and only care about fatal errors
+        let mut systeminfo_watcher = fuel_telemetry::systeminfo_watcher::SystemInfoWatcher::new();
+        if let Err(fuel_telemetry::errors::WatcherError::Fatal(_)) = systeminfo_watcher.start() {
```

Comment:
and here

Reply:
Yep. Implemented at the same time with ad0b3591.

Reply:
Seems to be addressed, closing the conversation.